Makefile

```diff
@@ -1,22 +1,94 @@
-PACKAGES := $(shell glide novendor)
+PKGS := $(shell go list ./... | grep -v go.uber.org/sally/vendor)
+SRCS := $(wildcard *.go)
 
-.PHONY: install
-install:
-	glide --version || go get github.com/Masterminds/glide
+all: test
+
+vendor-update:
+	rm -rf vendor
+	go get -d -v -t -u -f ./...
+	go get -v github.com/Masterminds/glide
+	glide create
+	glide update
+
+vendor-install:
+	go get -v github.com/Masterminds/glide
 	glide install
 
+build:
+	go build $(PKGS)
+
+install:
+	go install $(PKGS)
 
-.PHONY: lint
 lint:
-	go vet $(PACKAGES)
-	$(foreach pkg, $(PACKAGES), golint $(pkg) &&) echo "success"
+	go get -v github.com/golang/lint/golint
+	for file in $$(find . -name '*.go' | grep -v '^\./vendor'); do \
+		golint $$file; \
+		if [ -n "$$(golint $$file)" ]; then \
+			exit 1; \
+		fi; \
+	done
+
+vet:
+	go vet $(PKGS)
+
+errcheck:
+	go get -v github.com/kisielk/errcheck
+	errcheck $(PKGS)
+
+pretest: lint vet errcheck
+
+test: pretest
+	go test -race $(PKGS)
+
+clean:
+	go clean -i $(PKGS)
+	rm -rf _tmp
+
+docker-build-dev:
+	docker build -t uber/sally-dev .
+
+docker-test: docker-build-dev
+	docker run uber/sally-dev make test
+
+docker-build-internal:
+	rm -rf _tmp
+	mkdir -p _tmp
+	CGO_ENABLED=0 go build -a -installsuffix cgo -o _tmp/sally $(SRCS)
+	docker build -t uber/sally -f Dockerfile.sally .
+
+docker-build: docker-build-dev
+	docker run -v /var/run/docker.sock:/var/run/docker.sock uber/sally-dev make docker-build-internal
+
+docker-launch-dev-internal: install
+	sally
 
+docker-launch-dev: docker-build-dev
+	docker run -p 8080:8080 uber/sally-dev
 
-.PHONY: test
-test: lint
-	go test -race $(PACKAGES)
+docker-launch: docker-build
+	docker run -p 8080:8080 uber/sally
 
+launch: install
+	sally
 
-.PHONY: run
-run:
-	go build -i && ./sally
+.PHONY: \
```

Comment:
Why list all of the PHONY here instead of over each target? Seems like having them separate will only lead to forgetting to add when a new target is created.